### PR TITLE
[spark] supports schema merging when write

### PR DESCRIPTION
**Pull request by YannByron**
### Purpose

Linked issue: close https://github.com/apache/incubator-paimon/issues/1566

### CODE REVIEW

File: paimon-common/src/main/java/org/apache/paimon/types/ILength.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.types;
+
+/** An interface with a getLength method. */
+public interface ILength {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/types/ILength.java:22]: Do you think https://github.com/apache/incubator-paimon/pull/1685 works for you?

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+import org.apache.paimon.types.RowType;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/** The util class for merging the schemas. */
+public class SchemaMergingUtils {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java:40]: Maybe we can merge this class to `SchemaManager`?
And introduce a method to `mergeSchema(RowType rowType)` to `FileStore`?

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+     */
+    static DataType merge(DataType base0, DataType update0, AtomicInteger highestFieldId) {
+        // Here we try t0 merge the base0 and update0 without regard to the nullability,
+        // and set the base0's nullability to the return's.
+        DataType base = base0.copy(true);
+        DataType update = update0.copy(true);
+
+        if (base.equals(update)) {
+            return base0;
+        } else if (base instanceof RowType && update instanceof RowType) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java:85]: Maybe here can use `DataTypeDefaultVisitor`?

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        } else if (base instanceof DecimalType && update instanceof DecimalType) {
+            if (base.equals(update)) {
+                return base0;
+            } else {
+                throw new UnsupportedOperationException(
+                        String.format(
+                                "Failed to merge decimal types with different precision or scale: %s and %s",
+                                base, update));
+            }
+        } else if (DataTypeCasts.supportsImplicitCast(base, update)) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java:159]: Currently, we have using `supportsExplicitCast` to manually casting (From DDL alter table).
Do you think we should use `supportsExplicitCast`? Or provide an option to let user to choose?

**Review by JingsongLi — COMMENTED** [review]
Thanks @YannByron , Nice working!

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java [hunk 5]
```diff
@@ -818,6 +818,13 @@ public class CoreOptions implements Serializable {
                                     + " the value should be the user configured local time zone. The option value is either a full name"
                                     + " such as 'America/Los_Angeles', or a custom timezone id such as 'GMT-08:00'.");
 
+    public static final ConfigOption<Boolean> MERGE_SCHEMA =
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:821]: Maybe we can introduce a `SparkConnectorOptions`? This is only for spark.

Comment (YannByron) [reply paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:821]: I think so.

File: paimon-common/src/main/java/org/apache/paimon/types/ILength.java (hunk 1 above)
Comment (YannByron) [reply paimon-common/src/main/java/org/apache/paimon/types/ILength.java:22]: sure. very nice. I can't find it before.

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java (hunk 2 above)
Comment (YannByron) [reply paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java:40]: >> Maybe we can merge this class to SchemaManager?
Even though merge this to `SchemaManager`, all the methods will be `static`. i think there  is no different.

>> And introduce a method to mergeSchema(RowType rowType) to FileStore?
yes, better than the current `commitSchema`.

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java (hunk 4 above)
Comment (YannByron) [reply paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java:159]: that needs an option if we want to support Explicit-Casting.

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java (hunk 3 above)
Comment (YannByron) [reply paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java:85]: I think `DataTypeVisitor` is more used to extract some attributes, here I wanna know if a datatype is exactly what i want, and do some process if true. So maybe `instanceof` is clear enough.

Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java:85]: Sounds reasonable to me.

File: paimon-common/src/main/java/org/apache/paimon/types/DataField.java [hunk 6]
```diff
@@ -76,6 +77,11 @@ public DataType type() {
         return type;
     }
 
+    @VisibleForTesting
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/types/DataField.java:80]: A newId method looks good to me.
We can just remove `VisibleForTesting`. (Public class should not have public VisibleForTesting method)

File: paimon-core/src/main/java/org/apache/paimon/AbstractFileStore.java [hunk 7]
```diff
@@ -142,6 +142,11 @@ public CoreOptions options() {
         return options;
     }
 
+    @Override
+    public boolean mergeSchema(RowType rowType, Boolean allowExplicitCast) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/AbstractFileStore.java:146]: boolean instead of Boolean?

**Review by JingsongLi — COMMENTED** [review]
Thanks for the update, just left some minor comments.

File: paimon-core/src/main/java/org/apache/paimon/AbstractFileStore.java (hunk 7 above)
Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/AbstractFileStore.java:146]: All related call links use `boolean`. (Looks like you're confusing Java and scala?)

Comment (YannByron) [reply paimon-core/src/main/java/org/apache/paimon/AbstractFileStore.java:146]: yep, my fault.

**Review by JingsongLi — APPROVED** [review]
+1 Thanks @YannByron !

**Pull request merged**